### PR TITLE
refactor: return ExitCode from CliRunner::run()

### PR DESCRIPTION
**Pull request by samueltardieu**
`CliRunner::run()` has been renamed from `CliRunner::run_and_exit()` and returns an `ExitCode` which can be returned from `main()`.

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have added tests to cover my changes

### CODE REVIEW

**Comment by samueltardieu** [comment]
I think @yuja had something to say about it, according to #1071.

**Review by martinvonz — COMMENTED** [review]
> I think @yuja had something to say about it, according to #1071.

Yes, let's see what Yuya thinks.

File: CHANGELOG.md [hunk 1]
```diff
@@ -23,6 +23,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * `jj log` prefixes commit descriptions with "(empty)" when they contain no
   change compared to their parents.
 
+* The function `CliRunner::run_and_exit()` has been renamed `CliRunner::run()`
+  and returns a `std::process:ExitCode`.
+
```
[the remark this answers is not in the document]
Comment (samueltardieu) [reply CHANGELOG.md]: Removed.

**Comment by yuja** [comment]
In order to give caller control of process shutdown, I think it's better to not drop `ui` by callee.

```
let exit_code = run(&mut ui);
... do something that can fail, and print error to ui ...
ui.finalize_writes(); // wait pager
exit_code
```

That's what I meant in #1071. If `run() -> ExitCode` is just a nicer replacement of `run_and_exit() -> !`, this PR is fine.

**Comment by martinvonz** [comment]
> If `run() -> ExitCode` is just a nicer replacement of `run_and_exit() -> !`, this PR is fine.

Yes, that's what I think this PR is about. Just to be clear, are you saying that you're okay with the PR in its current form, and your concern was if it had looked different? Or do you still have some concern?

Since `run()` doesn't terminate after this PR, the caller can do more stuff after, but I can't think of any good reason they would want to. If we see a usecase for doing some late processing before the `Ui` instance is dropped, we can add more callbacks to `CliRunner`.

**Comment by yuja** [comment]
> > If `run() -> ExitCode` is just a nicer replacement of `run_and_exit() -> !`, this PR is fine.
> 
> Yes, that's what I think this PR is about. Just to be clear, are you saying that you're okay with the PR in its current form,

Yes. I'm okay with either form.

> and your concern was if it had looked different? Or do you still have some concern?

I have no other concern. I just don't think returning `ExitCode` is any better than `-> !` if the caller is supposed to do nothing after.

**Comment by martinvonz** [comment]
> I just don't think returning `ExitCode` is any better than `-> !` if the caller is supposed to do nothing after.

I agree with that. I think my TODO was just about using `ExitCode` for type safety and readability. However, I just found https://doc.rust-lang.org/std/process/fn.exit.html, which says:

> If a clean shutdown is needed it is recommended to only call this function at a known point where there are no more destructors left to run; or, preferably, simply return a type implementing [Termination](https://doc.rust-lang.org/std/process/trait.Termination.html) (such as [ExitCode](https://doc.rust-lang.org/std/process/struct.ExitCode.html) or Result) from the main function and avoid this function altogether

So I suppose they have a point there that about destructors.

**Review by martinvonz — APPROVED** [review]
Given the recommendation at https://doc.rust-lang.org/std/process/fn.exit.html, where they recommend returning an ExitCode (or similar), this LGTM.

**Pull request merged**